service/security/shiro/src/main/java/org/eclipse/kapua/service/authentication/shiro/realm/KapuaAuthenticatingRealm.java

```diff
@@ -235,19 +235,28 @@ protected void increaseLockoutPolicyCount(LoginAuthenticationInfo loginAuthentic
      * @since 2.0.0
      */
     protected void resetCredentialLockout(Credential credential) {
-        CredentialService credentialService = LOCATOR.getService(CredentialService.class);
-
-        credential.setFirstLoginFailure(null);
-        credential.setLoginFailuresReset(null);
-        credential.setLockoutReset(null);
-        credential.setLoginFailures(0);
-        try {
-            KapuaSecurityUtils.doPrivileged(() -> credentialService.update(credential));
-        } catch (KapuaException kex) {
-            throw new ShiroException("Unexpected error while looking for the lockout policy", kex);
+        //TODO find a proper way to update only if needed (obviously database update has a cost)
+        if (shouldResetCredentialLockout(credential)) {
```

Comment:
Shouldn't deciding whether to perform the update be the responsibility of the CredentialService - so that the rules are centralized there? Within the credential service there should still be time to decide whether the db update is needed or not

Reply:
the logic is already implemented here becasue if the login succeed then the reset lockout policy is executed. But if lockout policy is already reset this update is not needed. Unfortunately, if I'm right, the DAO layer doesn't perform any check if the entity had changed or not. I don't see any alternatives but, if you see one, please suggest me

Reply:
The DAO layer might not perform checks, but we still have a service layer we can act upon. And whether to invoke the DAO (repository, in the future) or not should be the responsibility of such service, IMHO - otherwise we risk having to replicate this logic elsewhere, should the need to change any of those field arise in the future.